lib/elixir/lib/kernel.ex

```diff
@@ -2642,6 +2642,126 @@ defmodule Kernel do
+
+  Raises an error if all conditions evaluate to `nil` or `false`.
+  For this reason, it may be necessary to add a final always-truthy condition
+  (anything non-`false` and non-`nil`), which will always match.
+
+  ## Examples
+
+      cond do
+        1 + 1 == 1 ->
+          "This will never match"
+        2 * 2 != 4 ->
+          "Nor this"
+        true ->
+          "This will"
+      end
+      #=> "This will"
+
+  """
+
+  defmacro cond(clauses) do
+    build_cond(clauses, __CALLER__)
+  end
+
+  defp build_cond([do: do_clause], caller) do
+    %{line: line} = caller
+
+    [{:->, meta, [[condition], body]} | tail] =
+      reversed_clauses = reverse_cond_clauses(do_clause, [])
+
+    {next_clauses, next_acc} =
+      case condition do
+        {:_, _, context} when is_atom(context) ->
+          raise ArgumentError,
+                "invalid use of _ inside \"cond\". If you want the last clause " <>
+                  "to always match, you probably meant to use: true ->"
+
+        value when is_atom(value) and value != false and value != nil ->
+          {tail, body}
+
+        _ ->
+          error = quote(do: raise(CondClauseError))
+          {reversed_clauses, error}
+      end
+
+    case_clause = build_cond_clauses(next_clauses, next_acc, meta)
+    replace_case_line(case_clause, line)
+  end
+
+  defp build_cond(_invalid_args, _caller) do
+    raise ArgumentError, "invalid arguments for \"cond\" as it expects a single :do keyword"
+  end
+
+  defp build_cond_clauses([], acc, _old_meta) do
+    acc
+  end
+
+  defp build_cond_clauses([{:->, new_meta, [[condition], body]} | tail], acc, old_meta) do
+    new_meta_line = get_line(new_meta)
+    old_meta_line = get_line(old_meta)
+
+    truthy_clause =
+      quote line: new_meta_line do
+        value when value != nil and value != false -> unquote(body)
+      end
+
+    falsy_clause =
+      quote line: old_meta_line do
+        _ -> unquote(acc)
+      end
+
+    case_clause =
+      quote line: new_meta_line do
+        case unquote(condition), do: unquote(truthy_clause ++ falsy_clause)
+      end
+
+    next_acc = optimize_boolean(case_clause, false)
+    build_cond_clauses(tail, next_acc, new_meta)
+  end
+
+  defp reverse_cond_clauses([], acc) do
+    acc
+  end
+
+  defp reverse_cond_clauses([{:->, _, [[_], _]} = head | tail], acc) do
+    reverse_cond_clauses(tail, [head | acc])
+  end
+
+  defp reverse_cond_clauses([{:->, _, [args, _]} | _], _acc) when is_list(args) do
+    raise ArgumentError, "expected one arg for :do clauses (->) in \"cond\""
+  end
+
+  defp reverse_cond_clauses(_invalid_clauses, _acc) do
+    raise ArgumentError, "expected -> clauses for :do in \"cond\""
+  end
+
+  defp get_line(meta) do
+    case :lists.keyfind(:line, 1, meta) do
+      {:line, line} -> line
+      false -> 0
+    end
+  end
+
+  defp replace_case_line({:case, meta, args}, line) do
+    new_meta = :lists.keyreplace(:line, 1, meta, {:line, line})
```

Comment:
I didn't know that `quote` can receive `line` option. I've changed in other places, but apparently, it doesn't work with pre-existing meta, with is strange for me:
```
quote line: 2 do
  unquote({:case, [line: 1], [true, [do: [{:->, [line: 1], [[true], true]}]]]})
end
```